File: go/layer2/layer2_test.go
package layer2

import (
	"errors"
	"strings"
	"testing"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// testSecret is a 32-byte HS256 secret used across all layer-2 tests.
// It is intentionally NOT a realistic looking value — the "test-" prefix
// makes it obvious the string is a fixture.
var testSecret = []byte("test-secret-1234567890123456789012")

func newTestClaims() ShortLivedClaims {
	return ShortLivedClaims{
		UID:             "user-1",
		DocumentName:    "doc-42",
		Role:            "editor",
		PermissionEpoch: 42,
	}
}

func TestIssueShortLivedTokenHappyPath(t *testing.T) {
	token, err := IssueShortLivedToken(IssueOptions{
		Claims: newTestClaims(),
		SigKey: testSecret,
		SigAlg: AlgHS256,
	})
	require.NoError(t, err)
	assert.NotEmpty(t, token)
	// JWT compact form has 3 dot-separated parts.
	assert.Equal(t, 3, strings.Count(token, ".")+1)
}

func TestIssueShortLivedTokenMissingUID(t *testing.T) {
	_, err := IssueShortLivedToken(IssueOptions{
		Claims: ShortLivedClaims{},
		SigKey: testSecret,
		SigAlg: AlgHS256,
	})
	require.Error(t, err)
	assert.ErrorIs(t, err, ErrMissingUID)
}

func TestIssueShortLivedTokenMissingSigKey(t *testing.T) {
	_, err := IssueShortLivedToken(IssueOptions{
		Claims: newTestClaims(),
		SigKey: nil,
		SigAlg: AlgHS256,
	})
	require.Error(t, err)
	assert.ErrorIs(t, err, ErrMissingSigKey)
}

func TestIssueShortLivedTokenShortHS256Key(t *testing.T) {
	_, err := IssueShortLivedToken(IssueOptions{
		Claims: newTestClaims(),
		SigKey: []byte("too-short-secret"),
		SigAlg: AlgHS256,
	})
	require.Error(t, err)
	assert.ErrorIs(t, err, ErrKeyTooShort)
}

func TestIssueShortLivedTokenEdDSAUnsupported(t *testing.T) {
	_, err := IssueShortLivedToken(IssueOptions{
		Claims: newTestClaims(),
		SigKey: testSecret,
		SigAlg: AlgEdDSA,
	})
	require.Error(t, err)
	assert.ErrorIs(t, err, ErrUnsupportedAlgorithm)
}

func TestIssueShortLivedTokenUnknownAlg(t *testing.T) {
	_, err := IssueShortLivedToken(IssueOptions{
		Claims: newTestClaims(),
		SigKey: testSecret,
		SigAlg: "RS256",
	})
	require.Error(t, err)
	assert.ErrorIs(t, err, ErrUnsupportedAlgorithm)
}

func TestVerifyShortLivedTokenRoundtrip(t *testing.T) {
	claims := newTestClaims()
	token, err := IssueShortLivedToken(IssueOptions{
		Claims: claims,
		SigKey: testSecret,
		SigAlg: AlgHS256,
	})
	require.NoError(t, err)

	got, err := VerifyShortLivedToken(token, VerifyOptions{
		SigKey: testSecret,
		SigAlg: AlgHS256,
	})
	require.NoError(t, err)
	require.NotNil(t, got)
	assert.Equal(t, claims.UID, got.UID)
	assert.Equal(t, claims.DocumentName, got.DocumentName)
	assert.Equal(t, claims.Role, got.Role)
	assert.Equal(t, claims.PermissionEpoch, got.PermissionEpoch)
	// ExpiresAt should be around now + DefaultTTL. Allow a broad window
	// so slow CI does not flake the assertion.
	assert.WithinDuration(t, time.Now().Add(DefaultTTL), got.ExpiresAt, 30*time.Second)
}

func TestVerifyShortLivedTokenCustomTTL(t *testing.T) {
	token, err := IssueShortLivedToken(IssueOptions{
		Claims: newTestClaims(),
		SigKey: testSecret,
		SigAlg: AlgHS256,
		TTL:    2 * time.Hour,
	})
	require.NoError(t, err)

	got, err := VerifyShortLivedToken(token, VerifyOptions{
		SigKey: testSecret,
		SigAlg: AlgHS256,
	})
	require.NoError(t, err)
	assert.WithinDuration(t, time.Now().Add(2*time.Hour), got.ExpiresAt, 30*time.Second)
}

func TestVerifyShortLivedTokenExpired(t *testing.T) {
	// Hand-build an already-expired JWT with the same secret.
	past := time.Now().Add(-1 * time.Hour)
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
		"uid":              "u-1",
		"document_name":    "d-1",
		"role":             "r-1",
		"permission_epoch": int64(1),
		"exp":              past.Unix(),
	})
	signed, err := token.SignedString(testSecret)
	require.NoError(t, err)

	_, verr := VerifyShortLivedToken(signed, VerifyOptions{
		SigKey: testSecret,
		SigAlg: AlgHS256,
	})
	require.Error(t, verr)
	assert.ErrorIs(t, verr, ErrTokenExpired)
}

func TestVerifyShortLivedTokenWrongSigKey(t *testing.T) {
	token, err := IssueShortLivedToken(IssueOptions{
		Claims: newTestClaims(),
		SigKey: testSecret,
		SigAlg: AlgHS256,
	})
	require.NoError(t, err)

	wrong := []byte("another-different-secret-1234567890")
	_, verr := VerifyShortLivedToken(token, VerifyOptions{
		SigKey: wrong,
		SigAlg: AlgHS256,
	})
	require.Error(t, verr)
	assert.ErrorIs(t, verr, ErrTokenInvalid)
}

func TestVerifyShortLivedTokenTampered(t *testing.T) {
	token, err := IssueShortLivedToken(IssueOptions{
		Claims: newTestClaims(),
		SigKey: testSecret,
		SigAlg: AlgHS256,
	})
	require.NoError(t, err)

	// Flip a character in the payload part.
	parts := strings.Split(token, ".")
	require.Len(t, parts, 3)
	// Replace one character in the payload's middle.
	payload := []byte(parts[1])
	payload[len(payload)/2] ^= 0x01
	tampered := parts[0] + "." + string(payload) + "." + parts[2]

	_, verr := VerifyShortLivedToken(tampered, VerifyOptions{
		SigKey: testSecret,
		SigAlg: AlgHS256,
	})
	require.Error(t, verr)
	assert.ErrorIs(t, verr, ErrTokenInvalid)
}

func TestVerifyShortLivedTokenMalformed(t *testing.T) {
	_, err := VerifyShortLivedToken("not-a-valid-jwt", VerifyOptions{
		SigKey: testSecret,
		SigAlg: AlgHS256,
	})
	require.Error(t, err)
	assert.ErrorIs(t, err, ErrTokenInvalid)
}

func TestVerifyShortLivedTokenMissingSigKey(t *testing.T) {
	_, err := VerifyShortLivedToken("x.y.z", VerifyOptions{
		SigKey: nil,
		SigAlg: AlgHS256,
	})
	require.Error(t, err)
	assert.ErrorIs(t, err, ErrMissingSigKey)
}

// TestVerifyShortLivedTokenShortHS256Key ensures the verify path rejects a
// weak HMAC secret symmetrically with the issue path. Without this guard, a
// verify-only consumer (token issued elsewhere) configured with a < 32-byte
// secret would silently accept the weak signature, enabling brute-force
// forgery of any uid/role/document_name/permission_epoch. Reviewer P1-A.
func TestVerifyShortLivedTokenShortHS256Key(t *testing.T) {
	// Sign a token with the same short secret so the signature would
	// otherwise verify — proves the guard fires before jwt.Parse.
	shortKey := []byte("too-short-secret")
	signed := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
		"uid": "u-1",
		"exp": time.Now().Add(time.Hour).Unix(),
	})
	tokenStr, err := signed.SignedString(shortKey)
	require.NoError(t, err)

	_, verr := VerifyShortLivedToken(tokenStr, VerifyOptions{
		SigKey: shortKey,
		SigAlg: AlgHS256,
	})
	require.Error(t, verr)
	assert.ErrorIs(t, verr, ErrKeyTooShort)
}

func TestVerifyShortLivedTokenEdDSAUnsupported(t *testing.T) {
	_, err := VerifyShortLivedToken("x.y.z", VerifyOptions{
		SigKey: testSecret,
		SigAlg: AlgEdDSA,
	})
	require.Error(t, err)
	assert.ErrorIs(t, err, ErrUnsupportedAlgorithm)
}

func TestVerifyShortLivedTokenMissingUIDClaim(t *testing.T) {
	// Sign a token that has no uid claim.
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
		"document_name": "d",
		"exp":           time.Now().Add(time.Hour).Unix(),
	})
	signed, err := token.SignedString(testSecret)
	require.NoError(t, err)

	_, verr := VerifyShortLivedToken(signed, VerifyOptions{
		SigKey: testSecret,
		SigAlg: AlgHS256,
	})
	require.Error(t, verr)
	assert.ErrorIs(t, verr, ErrTokenInvalid)
}

// TestVerifyShortLivedTokenAlgorithmConfusion covers the JWT alg-confusion
// class of attacks: a token signed with a different-family algorithm (RS256,
// none, ...) must be rejected even if the alg header looks legitimate.
func TestVerifyShortLivedTokenNoneAlgRejected(t *testing.T) {
	// Manually craft an "alg: none" token; golang-jwt/v5 refuses to sign
	// with unsafe methods via NewWithClaims + SignedString, but we can
	// still forge the compact form by hand.
	// header {"alg":"none","typ":"JWT"} base64url = eyJhbGciOiJub25lIiwidHlwIjoiSldUIn0
	// payload {"uid":"u-1","exp":9999999999} = eyJ1aWQiOiJ1LTEiLCJleHAiOjk5OTk5OTk5OTl9
	// signature = ""
	forged := "eyJhbGciOiJub25lIiwidHlwIjoiSldUIn0.eyJ1aWQiOiJ1LTEiLCJleHAiOjk5OTk5OTk5OTl9."
	_, err := VerifyShortLivedToken(forged, VerifyOptions{
		SigKey: testSecret,
		SigAlg: AlgHS256,
	})
	require.Error(t, err)
	assert.ErrorIs(t, err, ErrTokenInvalid)
}

func TestSentinelErrorsUnwrappable(t *testing.T) {
	// Guard: newer callers use errors.Is against sentinels. Verify the
	// core sentinels are directly comparable and not shadowed by wrapping.
	assert.True(t, errors.Is(ErrMissingUID, ErrMissingUID))
	assert.True(t, errors.Is(ErrMissingSigKey, ErrMissingSigKey))
	assert.True(t, errors.Is(ErrTokenExpired, ErrTokenExpired))
	assert.True(t, errors.Is(ErrTokenInvalid, ErrTokenInvalid))
	assert.True(t, errors.Is(ErrUnsupportedAlgorithm, ErrUnsupportedAlgorithm))
	assert.True(t, errors.Is(ErrKeyTooShort, ErrKeyTooShort))
}

func TestShortLivedClaimsFromMapPreservesEpoch(t *testing.T) {
	// Directly test the int64 preservation code path since the JWT lib
	// naturally rounds through float64.
	m := jwt.MapClaims{
		"uid":              "u-1",
		"permission_epoch": float64(1_700_000_000),
		"exp":              float64(time.Now().Add(time.Hour).Unix()),
	}
	got, err := shortLivedClaimsFromMap(m)
	require.NoError(t, err)
	assert.Equal(t, int64(1_700_000_000), got.PermissionEpoch)
	assert.False(t, got.ExpiresAt.IsZero())
}

func TestShortLivedClaimsFromMapAcceptsIntEpoch(t *testing.T) {
	m := jwt.MapClaims{
		"uid":              "u-1",
		"permission_epoch": int64(7),
	}
	got, err := shortLivedClaimsFromMap(m)
	require.NoError(t, err)
	assert.Equal(t, int64(7), got.PermissionEpoch)
}

func TestShortLivedClaimsFromMapAcceptsPlainIntEpoch(t *testing.T) {
	m := jwt.MapClaims{
		"uid":              "u-1",
		"permission_epoch": int(9),
	}
	got, err := shortLivedClaimsFromMap(m)
	require.NoError(t, err)
	assert.Equal(t, int64(9), got.PermissionEpoch)
}

func TestShortLivedClaimsFromMapAcceptsInt64Exp(t *testing.T) {
	future := time.Now().Add(time.Hour).Unix()
	m := jwt.MapClaims{
		"uid": "u-1",
		"exp": future,
	}
	got, err := shortLivedClaimsFromMap(m)
	require.NoError(t, err)
	assert.Equal(t, time.Unix(future, 0).UTC(), got.ExpiresAt)
}

func TestShortLivedClaimsFromMapDropsUnexpectedEpochType(t *testing.T) {
	m := jwt.MapClaims{
		"uid":              "u-1",
		"permission_epoch": "not-a-number", // gets ignored per type switch
	}
	got, err := shortLivedClaimsFromMap(m)
	require.NoError(t, err)
	assert.Equal(t, int64(0), got.PermissionEpoch)
}

// TestVerifyShortLivedTokenRejectsHS512 verifies that a token signed with a
// different HMAC family (HS512) is rejected — the WithValidMethods filter
// must reject it before the key callback even fires.
func TestVerifyShortLivedTokenRejectsHS512(t *testing.T) {
	token := jwt.NewWithClaims(jwt.SigningMethodHS512, jwt.MapClaims{
		"uid": "u-1",
		"exp": time.Now().Add(time.Hour).Unix(),
	})
	signed, err := token.SignedString(testSecret)
	require.NoError(t, err)

	_, verr := VerifyShortLivedToken(signed, VerifyOptions{
		SigKey: testSecret,
		SigAlg: AlgHS256,
	})
	require.Error(t, verr)
	assert.ErrorIs(t, verr, ErrTokenInvalid)
}

File: ts/tests/layer2.test.ts
import { describe, expect, it } from 'vitest'
import { SignJWT } from 'jose'

import { OctoAuthError } from '../src/index.js'
import {
  AlgEdDSA,
  AlgHS256,
  DEFAULT_TTL_SECONDS,
  MIN_HS256_KEY_SIZE,
  issueShortLivedToken,
  verifyShortLivedToken,
} from '../src/layer2.js'

const goodKey = new Uint8Array(32).fill(9)

describe('layer2 — HS256 round-trip', () => {
  it('issues and verifies a valid token', async () => {
    const token = await issueShortLivedToken({
      claims: {
        uid: 'u1',
        documentName: 'doc-1',
        role: 'writer',
        permissionEpoch: 3,
      },
      sigKey: goodKey,
      sigAlg: AlgHS256,
      ttlSeconds: 60,
    })
    const claims = await verifyShortLivedToken(token, {
      sigKey: goodKey,
      sigAlg: AlgHS256,
    })
    expect(claims.uid).toBe('u1')
    expect(claims.documentName).toBe('doc-1')
    expect(claims.role).toBe('writer')
    expect(claims.permissionEpoch).toBe(3)
    expect(claims.exp).toBeGreaterThan(Math.floor(Date.now() / 1000))
  })

  it('default TTL applies when ttlSeconds unset', async () => {
    expect(DEFAULT_TTL_SECONDS).toBe(300)
    const before = Math.floor(Date.now() / 1000)
    const token = await issueShortLivedToken({
      claims: { uid: 'u', documentName: '', role: '', permissionEpoch: 0 },
      sigKey: goodKey,
      sigAlg: AlgHS256,
    })
    const c = await verifyShortLivedToken(token, {
      sigKey: goodKey,
      sigAlg: AlgHS256,
    })
    // exp should be roughly before + DEFAULT_TTL_SECONDS (±2s for clock).
    expect(c.exp).toBeGreaterThanOrEqual(before + DEFAULT_TTL_SECONDS - 2)
    expect(c.exp).toBeLessThanOrEqual(before + DEFAULT_TTL_SECONDS + 2)
  })
})

describe('layer2 — issue validation', () => {
  it('rejects missing uid', async () => {
    await expect(
      issueShortLivedToken({
        claims: { uid: '', documentName: '', role: '', permissionEpoch: 0 },
        sigKey: goodKey,
        sigAlg: AlgHS256,
      }),
    ).rejects.toMatchObject({ kind: 'invalid-credential' })
  })

  it('rejects empty sigKey', async () => {
    await expect(
      issueShortLivedToken({
        claims: { uid: 'u', documentName: '', role: '', permissionEpoch: 0 },
        sigKey: new Uint8Array(0),
        sigAlg: AlgHS256,
      }),
    ).rejects.toMatchObject({ kind: 'invalid-credential' })
  })

  it('rejects short HS256 sigKey (< 32 bytes)', async () => {
    await expect(
      issueShortLivedToken({
        claims: { uid: 'u', documentName: '', role: '', permissionEpoch: 0 },
        sigKey: new Uint8Array(MIN_HS256_KEY_SIZE - 1),
        sigAlg: AlgHS256,
      }),
    ).rejects.toThrow(/HS256 secret must be at least/)
  })

  it('rejects EdDSA (v1 gap → unsupported-algorithm)', async () => {
    try {
      await issueShortLivedToken({
        claims: { uid: 'u', documentName: '', role: '', permissionEpoch: 0 },
        sigKey: goodKey,
        sigAlg: AlgEdDSA,
      })
      expect.fail('should have thrown')
    } catch (err) {
      expect(err).toBeInstanceOf(OctoAuthError)
      expect((err as OctoAuthError).kind).toBe('infra-failure')
      expect((err as Error).message).toMatch(/unsupported signature algorithm/)
    }
  })

  it('rejects unknown algorithm', async () => {
    await expect(
      issueShortLivedToken({
        claims: { uid: 'u', documentName: '', role: '', permissionEpoch: 0 },
        sigKey: goodKey,
        sigAlg: 'RS256' as unknown as 'HS256',
      }),
    ).rejects.toThrow(/unsupported signature algorithm/)
  })
})

describe('layer2 — verify validation', () => {
  it('rejects empty sigKey', async () => {
    await expect(
      verifyShortLivedToken('anything', {
        sigKey: new Uint8Array(0),
        sigAlg: AlgHS256,
      }),
    ).rejects.toMatchObject({ kind: 'invalid-credential' })
  })

  it('rejects EdDSA at verify time (v1 gap)', async () => {
    await expect(
      verifyShortLivedToken('anything', {
        sigKey: goodKey,
        sigAlg: AlgEdDSA,
      }),
    ).rejects.toMatchObject({ kind: 'infra-failure' })
  })

  it('rejects unknown algorithm at verify time', async () => {
    await expect(
      verifyShortLivedToken('anything', {
        sigKey: goodKey,
        sigAlg: 'RS256' as unknown as 'HS256',
      }),
    ).rejects.toThrow(/unsupported signature algorithm/)
  })

  it('rejects malformed token', async () => {
    await expect(
      verifyShortLivedToken('not.a.jwt', { sigKey: goodKey, sigAlg: AlgHS256 }),
    ).rejects.toMatchObject({ kind: 'invalid-credential' })
  })
})

describe('layer2 — security invariants', () => {
  it('rejects wrong-key signature', async () => {
    const token = await issueShortLivedToken({
      claims: { uid: 'u', documentName: '', role: '', permissionEpoch: 0 },
      sigKey: goodKey,
      sigAlg: AlgHS256,
    })
    await expect(
      verifyShortLivedToken(token, {
        sigKey: new Uint8Array(32).fill(0),
        sigAlg: AlgHS256,
      }),
    ).rejects.toMatchObject({ kind: 'invalid-credential' })
  })

  it('rejects tampered payload', async () => {
    const token = await issueShortLivedToken({
      claims: { uid: 'u', documentName: '', role: '', permissionEpoch: 0 },
      sigKey: goodKey,
      sigAlg: AlgHS256,
    })
    const parts = token.split('.')
    // Flip a char in the payload segment; signature no longer matches.
    const bad = parts[1]!.slice(0, -1) + (parts[1]!.slice(-1) === 'A' ? 'B' : 'A')
    const tampered = `${parts[0]}.${bad}.${parts[2]}`
    await expect(
      verifyShortLivedToken(tampered, { sigKey: goodKey, sigAlg: AlgHS256 }),
    ).rejects.toMatchObject({ kind: 'invalid-credential' })
  })

  it('rejects expired tokens', async () => {
    // Sign with a manual jose call to get a past exp — issueShortLivedToken
    // requires ttlSeconds > 0.
    const nowSec = Math.floor(Date.now() / 1000)
    const past = nowSec - 60
    const token = await new SignJWT({ uid: 'u' })
      .setProtectedHeader({ alg: 'HS256' })
      .setExpirationTime(past)
      .sign(goodKey)
    await expect(
      verifyShortLivedToken(token, { sigKey: goodKey, sigAlg: AlgHS256 }),
    ).rejects.toMatchObject({ kind: 'invalid-credential' })
  })

  it('rejects HS512 tokens (alg-confusion defense)', async () => {
    // Craft a HS512 token that the verifier should reject even though the
    // secret material is the same length. jose's `algorithms: ['HS256']`
    // whitelist enforces this.
    const bigKey = new Uint8Array(64).fill(9)
    const token = await new SignJWT({ uid: 'u' })
      .setProtectedHeader({ alg: 'HS512' })
      .setExpirationTime('5m')
      .sign(bigKey)
    await expect(
      verifyShortLivedToken(token, { sigKey: bigKey, sigAlg: AlgHS256 }),
    ).rejects.toMatchObject({ kind: 'invalid-credential' })
  })

  it('rejects alg=none tokens', async () => {
    // Header + body base64url without a signature; jose refuses this.
    const header = Buffer.from('{"alg":"none"}').toString('base64url')
    const payload = Buffer.from(
      JSON.stringify({ uid: 'u', exp: Math.floor(Date.now() / 1000) + 300 }),
    ).toString('base64url')
    const evil = `${header}.${payload}.`
    await expect(
      verifyShortLivedToken(evil, { sigKey: goodKey, sigAlg: AlgHS256 }),
    ).rejects.toMatchObject({ kind: 'invalid-credential' })
  })

  it('rejects tokens missing uid claim', async () => {
    // Issue a token with an empty uid via jose directly (issueShortLivedToken
    // would refuse).
    const token = await new SignJWT({ document_name: 'd' })
      .setProtectedHeader({ alg: 'HS256' })
      .setExpirationTime('5m')
      .sign(goodKey)
    await expect(
      verifyShortLivedToken(token, { sigKey: goodKey, sigAlg: AlgHS256 }),
    ).rejects.toMatchObject({ kind: 'invalid-credential' })
  })
})
